pkg/install/drbd/drbd_adapter.go

```diff
@@ -98,375 +104,537 @@ func HandelDRBDConfigs(clusterInstance *hwameistoriov1alpha1.Cluster) {
 	nodeAffinity = *drbdConfigs.NodeAffinity
 }
 
-func CreateDRBDAdapter(cli client.Client) (int, error) {
-	nodeList := corev1.NodeList{}
-	if err := cli.List(context.TODO(), &nodeList); err != nil {
-		log.Errorf("List nodes err: %v", err)
-		return adapterCreatedJobNum, err
+func FulfillDRBDSpec(clusterInstance *hwameistoriov1alpha1.Cluster) *hwameistoriov1alpha1.Cluster {
+	if clusterInstance.Spec.DRBD == nil {
+		clusterInstance.Spec.DRBD = &hwameistoriov1alpha1.DRBDSpec{}
+	}
+	if clusterInstance.Spec.DRBD.DeployOnMaster == "" {
+		clusterInstance.Spec.DRBD.DeployOnMaster = defaultDeployOnMaster
 	}
 
-	for _, node := range nodeList.Items {
-		_, masterLabelExist := node.Labels["node-role.kubernetes.io/master"]
-		_, controlPlaneLabel := node.Labels["node-role.kubernetes.io/control-plane"]
-		if masterLabelExist || controlPlaneLabel {
-			if !deployOnMaster {
-				continue
-			}
+	if clusterInstance.Spec.DRBD.Shipper == nil {
+		clusterInstance.Spec.DRBD.Shipper = &operatorv1alpha1.ImageSpec{}
+	}
+	if clusterInstance.Spec.DRBD.Shipper.Registry == "" {
+		clusterInstance.Spec.DRBD.Shipper.Registry = defaultImageRegistry
+	}
+	if clusterInstance.Spec.DRBD.Shipper.Repository == "" {
+		clusterInstance.Spec.DRBD.Shipper.Repository = defaultShipperRepository
+	}
+
+	if clusterInstance.Spec.DRBD.Shipper.Tag == "" {
+		clusterInstance.Spec.DRBD.Shipper.Tag = defaultShipperChar
+	}
+
+	if clusterInstance.Spec.DRBD.DRBDVersion == "" {
+		clusterInstance.Spec.DRBD.DRBDVersion = defaultDRBDVersion
+	}
+	if clusterInstance.Spec.DRBD.Upgrade == "" {
+		clusterInstance.Spec.DRBD.Upgrade = defaultDRBDUpgrade
+	}
+	if clusterInstance.Spec.DRBD.CheckHostName == "" {
+		clusterInstance.Spec.DRBD.CheckHostName = defaultCheckHostName
+	}
+	if clusterInstance.Spec.DRBD.UseAffinity == "" {
+		clusterInstance.Spec.DRBD.UseAffinity = defaultUseAffinity
+	}
+	if clusterInstance.Spec.DRBD.NodeAffinity == nil {
+		clusterInstance.Spec.DRBD.NodeAffinity = &corev1.NodeAffinity{}
+	}
+	if clusterInstance.Spec.DRBD.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		clusterInstance.Spec.DRBD.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = &corev1.NodeSelector{}
+	}
+	if clusterInstance.Spec.DRBD.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms == nil {
+		clusterInstance.Spec.DRBD.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms = defaultNodeSelectTerms
+	}
+	if clusterInstance.Spec.DRBD.ChartVersion == "" {
+		clusterInstance.Spec.DRBD.ChartVersion = defaultChartVersion
+	}
+
+	return clusterInstance
+}
+
+func CreateDRBDJob(cli client.Client, node corev1.Node) error {
+
+	_, masterLabelExist := node.Labels["node-role.kubernetes.io/master"]
+	_, controlPlaneLabel := node.Labels["node-role.kubernetes.io/control-plane"]
+	if masterLabelExist || controlPlaneLabel {
+		if !deployOnMaster {
+			return nil
 		}
+	}
```

Comment:
can we apply these 2 labels as selector before create drbd job on each node?

Reply:
OK, I have simplified the process and efficiency. Please review it again when you are free.